### PR TITLE
[Sema] add cast from IncompleteArrayType to ConstantArrayType in TryReferenceListInitialization

### LINKED ISSUE #62945: clang crash in CodeGen
This code can be compiled by gcc but crash clang

```c++
void f(int (&&)[]);
void foo(int a) { f({a}); }
```

https://godbolt.org/z/3qb6x3Y48

error stack

Assertion:

```console
clang++: /root/llvm-project/clang/lib/CodeGen/CGCall.cpp:4372: 
void clang::CodeGen::CodeGenFunction::EmitCallArgs(clang::CodeGen::CallArgList&, clang::CodeGen::CodeGenFunction::PrototypeWrapper, llvm::iterator_range<clang::Stmt::CastIterator<clang::Expr, const clang::Expr* const, const clang::Stmt* const> >, clang::CodeGen::CodeGenFunction::AbstractCallee, unsigned int, clang::CodeGen::CodeGenFunction::EvaluationOrder): 
Assertion `(isGenericMethod || Ty->isVariablyModifiedType() || Ty.getNonReferenceType()->isObjCRetainableType() || getContext() .getCanonicalType(Ty.getNonReferenceType()) .getTypePtr() == getContext().getCanonicalType((*Arg)->getType()).getTypePtr()) && "type mismatch in call argument!"' failed.
```

Backtrace
```console
8  clang-17                 0x0000000104b9fe48 clang::CodeGen::CodeGenFunction::EmitCallArgs(clang::CodeGen::CallArgList&, clang::CodeGen::CodeGenFunction::PrototypeWrapper, llvm::iterator_range<clang::Stmt::CastIterator<clang::Expr, clang::Expr const* const, clang::Stmt const* const>>, clang::CodeGen::CodeGenFunction::AbstractCallee, unsigned int, clang::CodeGen::CodeGenFunction::EvaluationOrder) + 996
9  clang-17                 0x0000000104dab534 clang::CodeGen::CodeGenFunction::EmitCall(clang::QualType, clang::CodeGen::CGCallee const&, clang::CallExpr const*, clang::CodeGen::ReturnValueSlot, llvm::Value*) + 3004
10 clang-17                 0x0000000104daa3e4 clang::CodeGen::CodeGenFunction::EmitCallExpr(clang::CallExpr const*, clang::CodeGen::ReturnValueSlot) + 632
11 clang-17                 0x0000000104e23130 (anonymous namespace)::ScalarExprEmitter::VisitCallExpr(clang::CallExpr const*) + 176
12 clang-17                 0x0000000104e1d08c clang::StmtVisitorBase<std::__1::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) + 4172
13 clang-17                 0x0000000104e13ea0 (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) + 60
14 clang-17                 0x0000000104e24384 (anonymous namespace)::ScalarExprEmitter::VisitExprWithCleanups(clang::ExprWithCleanups*) + 80
15 clang-17                 0x0000000104e1d30c clang::StmtVisitorBase<std::__1::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) + 4812
16 clang-17                 0x0000000104e13ea0 (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) + 60
17 clang-17                 0x0000000104e13de8 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) + 204
18 clang-17                 0x0000000104d8ce9c clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) + 148
19 clang-17                 0x0000000104d8cd68 clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) + 108
20 clang-17                 0x0000000104fd2984 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) + 632
21 clang-17                 0x0000000104fdcbf4 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) + 844
22 clang-17                 0x00000001050a37d0 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) + 128
23 clang-17                 0x00000001050a4420 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) + 1648
24 clang-17                 0x00000001050d9c5c clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) + 476
25 clang-17                 0x00000001050cfc74 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) + 648
26 clang-17                 0x00000001050d52dc clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) + 1748
27 clang-17                 0x00000001050ce318 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) + 232
28 clang-17                 0x00000001052eb520 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) + 164

```

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixed: https://github.com/llvm/llvm-project/issues/62945
c++20 supports "Permit conversions to arrays of unknown bound". This need additional cast from IncompleteArrayType to ConstantArrayType in TryReferenceListInitialization

### CODE REVIEW

**Comment by HerrCai0907** [comment]
from https://reviews.llvm.org/D151515

**Comment by cor3ntin** [comment]
@Fznamznon

File: clang/lib/Sema/SemaInit.cpp [hunk 1]
```diff
@@ -4532,6 +4532,15 @@ static void TryReferenceListInitialization(Sema &S,
...
         Sequence.AddQualificationConversionStep(
             cv1T1, DestType->isRValueReferenceType() ? VK_XValue : VK_LValue);
+      else if (S.getLangOpts().CPlusPlus20 &&
+               isa<IncompleteArrayType>(T1->getUnqualifiedDesugaredType()) &&
+               DestType->isRValueReferenceType()) {
+        // [dcl.init.list] p3.10
+        // unless T is “reference to array of unknown bound of U”, in which case
+        // the type of the prvalue is the type of x in the declaration U x[] H,
+        // where H is the initializer list.
+        Sequence.AddQualificationConversionStep(cv1T1, VK_XValue);
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaInit.cpp:4542]: Is qualification conversion right here? I'm not sure we convert qualifiers here.

File: clang/lib/Sema/SemaInit.cpp [hunk 2]
```diff
@@ -4532,6 +4532,15 @@ static void TryReferenceListInitialization(Sema &S,
       if (T1Quals.hasAddressSpace())
         Sequence.AddQualificationConversionStep(
             cv1T1, DestType->isRValueReferenceType() ? VK_XValue : VK_LValue);
+      else if (S.getLangOpts().CPlusPlus20 &&
+               isa<IncompleteArrayType>(T1->getUnqualifiedDesugaredType()) &&
+               DestType->isRValueReferenceType()) {
+        // [dcl.init.list] p3.10
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaInit.cpp:4538]: NIT
```suggestion
        // C++20 [dcl.init.list]p3.10:
```

**Review by Fznamznon — COMMENTED** [review]
This needs a release note.

cc @erichkeane

File: clang/lib/Sema/SemaInit.cpp (hunk 1 above)
Comment (HerrCai0907) [reply clang/lib/Sema/SemaInit.cpp]: I have changed sequence in the latest commit.
The reason to qualification conversion is that in cpp 20, spec. requires support cast from `initializer list` to `reference to array of unknown bound`.

for example: 
```
void foo(int a) {
  auto f = [](int(&&)[]) {};
  f({a});
}
```

`{a}` should be casted from int[1] to int[].


ast will look like:
```
      `-CXXOperatorCallExpr 0x12b904d18 <col:3, col:8> 'void':'void' '()'
        |-ImplicitCastExpr 0x12b904c58 <col:4, col:8> 'void (*)(int (&&)[]) const' <FunctionToPointerDecay>
        | `-DeclRefExpr 0x12b904bd8 <col:4, col:8> 'void (int (&&)[]) const' lvalue CXXMethod 0x12b904008 'operator()' 'void (int (&&)[]) const'
        |-ImplicitCastExpr 0x12b904c70 <col:3> 'const (lambda at ID/test.cpp:2:12)' lvalue <NoOp>
        | `-DeclRefExpr 0x12b904b10 <col:3> '(lambda at ID/test.cpp:2:12)':'(lambda at ID/test.cpp:2:12)' lvalue Var 0x12b8ee210 'f' '(lambda at ID/test.cpp:2:12)':'(lambda at ID/test.cpp:2:12)'
        `-MaterializeTemporaryExpr 0x12b904d00 <col:5, col:7> 'int[]':'int[]' xvalue
          `-ImplicitCastExpr 0x12b904ce8 <col:5, col:7> 'int[]':'int[]' <NoOp>
            `-InitListExpr 0x12b904c88 <col:5, col:7> 'int[1]'
              `-ImplicitCastExpr 0x12b904cc8 <col:6> 'int' <LValueToRValue>
                `-DeclRefExpr 0x12b904b30 <col:6> 'int' lvalue ParmVar 0x12b8edfb8 'a' 'int'
```

Comment (Fznamznon) [reply clang/lib/Sema/SemaInit.cpp]: @HerrCai0907 I see what you mean and what the patch accomplishes. But my confusion was, why in order to do that `AddQualificationConversionStep` is called? IMO this is just to convert cv-qualifiers. The goal is to convert one array type to another, right?

Comment (HerrCai0907) [reply clang/lib/Sema/SemaInit.cpp]: Yes, int[1] to int[] in above example

**Comment by efriedma-quic** [comment]
I guess this ends up producing a CK_NoOp CastExpr?  That's probably okay, but can we amend the documentation for CK_NoOp to give this as an example?

**Comment by HerrCai0907** [comment]
> I guess this ends up producing a CK_NoOp CastExpr? That's probably okay, but can we amend the documentation for CK_NoOp to give this as an example?

Done

**Comment by efriedma-quic** [comment]
LGTM

**Comment by shafik** [comment]
This is linked to the following regression: https://github.com/llvm/llvm-project/issues/151716

**Pull request merged**